internal/provider/notification_recipients.go

```diff
@@ -84,6 +85,33 @@ func notificationRecipientSchema(allowedTypes []client.RecipientType) schema.Set
 	}
 }
 
+func reconcileNotificationRecipientState(remote []client.NotificationRecipient, state []models.NotificationRecipientModel) []models.NotificationRecipientModel {
+	if state == nil {
+		// if we don't have any state, we can't reconcile anything so just return the remote recipients
+		return flattenNotificationRecipients(remote)
+	}
+
+	recipients := make([]models.NotificationRecipientModel, len(remote))
+	// match the remote recipients to those in the state sorting out type+target vs ID
+	for i, r := range remote {
+		idx := slices.IndexFunc(state, func(s models.NotificationRecipientModel) bool {
+			if !s.ID.IsNull() {
+				return s.ID.ValueString() == r.ID
+			}
+			return s.Type.ValueString() == string(r.Type) && s.Target.ValueString() == r.Target
+		})
+		if idx < 0 {
+			// if we didn't find a match, use the recipient as specified in remote
+			recipients[i] = notificationRecipientToModel(r)
+		} else {
+			// if we found a match, use the stored recipient
+			recipients[i] = state[idx]
```

Comment:
Are `state` and `remote` guaranteed to be of the same length?

Reply:
Ah, good catch! They are not: I'll put in a check

Reply:
I think I was misreading index variables here. This is fine 👍